pkg/intermediate/aggregate.go

```diff
@@ -0,0 +1,284 @@
+package intermediate
+
+import (
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/vmware/go-ipfix/pkg/entities"
+	"github.com/vmware/go-ipfix/pkg/registry"
+)
+
+type aggregation struct {
+	// tupleRecordMap maps each connection with its records
+	tupleRecordMap map[Tuple][]entities.Record
+	// tupleRecordLock allows multiple readers or one writer at the same time
+	tupleRecordLock sync.RWMutex
+	// workerPool is for storing worker channels to process the messages
+	workerPool chan chan *entities.Message
```

Comment:
Is that possible to define a single channel shared by all workers,
i.e. all workers consume messages from the same channel and every worker can get a message from the channel when it is available?

Reply:
The reason of passing channels over channels is that it is sending message(job) to available worker and make sure each worker is processing one message(job) at a time. The worker pool contains the channels that are available to receive message.

Reply:
Thanks for the explanation. I did a simple test. It looks like a single channel shared by multiple workers can achieve the same goal. For example,

```
package main

import (
        "fmt"
        "time"
)

const NUM_WORKERS int = 3

func produce(jobs chan<- int) {
	for i := 1; i <= 10; i++ {
                jobs <- i
	}
	close(jobs)
}

func consume(worker int, jobs <-chan int, done chan<- bool) {
        for job := range jobs {
                fmt.Printf("[%v] worker %d got %d\n", time.Now().Format(time.Stamp), worker, job)
                time.Sleep(time.Duration(worker) * time.Second)
        }
        done <- true
}

func main() {
        jobs := make(chan int)
	done := make(chan bool)

        go produce(jobs)

	for i := 1; i <= NUM_WORKERS; i++ {
		go consume(i, jobs, done)
	}

        for i := 1; i <= NUM_WORKERS; i++ {
		<-done
	}
}
```

Then the output is like:

$ go run  test.go
[Nov 11 20:57:03] worker 3 got 1
[Nov 11 20:57:03] worker 1 got 2
[Nov 11 20:57:03] worker 2 got 3
[Nov 11 20:57:04] worker 1 got 4
[Nov 11 20:57:05] worker 2 got 5
[Nov 11 20:57:05] worker 1 got 6
[Nov 11 20:57:06] worker 3 got 7
[Nov 11 20:57:06] worker 1 got 8
[Nov 11 20:57:07] worker 2 got 9
[Nov 11 20:57:07] worker 1 got 10

Reply:
Yes, it can also achieve that. The way we did for intermediate process is that we create several workers (each is a goroutine with a channel). then we put the channel into the worker pool for producers to receive message. The channel (worker) which gets the message will start processing it and then notify the worker pool by trying to sending the channel back to worker pool. It mimics the way of sending jobs to go routine and this model is used by similar project [here](https://github.com/cloudflare/goflow/blob/efa7272e2cc3aa91c1d904e982b4f8a8eba28460/decoders/decoder.go) and some [examples](https://medium.com/better-programming/hands-on-go-concurrency-the-producer-consumer-pattern-c42aab4e3bd2) for comparison.
Actually I don't see one is much better than the other. The advantage I can see of using passing channel over channel is that only available worker will be put into the worker pool to process the message and the finish of each job can be notified.

Reply:
Thanks for the details. My example is actually copied from the Single Producer Multiple Consumer case in the same [examples](https://medium.com/better-programming/hands-on-go-concurrency-the-producer-consumer-pattern-c42aab4e3bd2) link. So I was curious why we require channel over channel in our scenario.
Sorry one more question. When you said "the finish of each job can be notified" is an advantage of using channel over channel, did you just mean this can be done with channel over channel, but it is not implemented this way in our code, Is that right? because I do not see we ack the finish of job in the channel over channel. Thanks again!

Reply:
The suggestion makes sense to me. And it is easier to understand if not using channel over channel. will try to modify the code in a separate PR. Thanks!

Reply:
Hi Shihao,
Went through this thread in a more detailed way. The approach of using one channel sounds good. However, I have a question related to the code snippet provided here: https://github.com/vmware/go-ipfix/pull/51#discussion_r521841193

Will all consume go routines exit properly? A bool signal on unbuffered channel `done` sent by say consume-1, and it is received by main, then main exits and consume-1 exits too. What about consume-2 and consume-3 which are in blocked state after sending on `done`. As there are no receivers for these signals, they will not exit properly. Do you agree?

Reply:
Yes, the complete code should be doing
```
        for i := 1; i <= NUM_WORKERS; i++ {
		<-done
	}
```
I was too lazy :-) Thanks.